keps/sig-cli/2368-kubectl-default-container/README.md

```diff
@@ -0,0 +1,277 @@
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Introduce an optional annotation for consumption by certain kubectl commands which will allow picking a default container.
+
+## Motivation
+
+
+As the default log annotation is added, the motivation of it is like below
+> Kubectl provides a number of commands to simplify working with Kubernetes by making requests to
+> Resources and SubResources.  These requests are mostly static, with fields filled in by user
+> supplied flags.  Today the commands are compiled into the client, which as the following challenges:
+
+Sidecars are popular now.
+Alternatively, a way to generically describe the default container for anything that happens to need a "default", which could also be used by external tooling.
+
+Issues https://github.com/kubernetes/kubernetes/issues/96986 opened by @howardjohn 
+> However, it gets worse because aside from a warning, this couples the default exec container to the container order. 
+> The container ordering also happens to have an impact on container startup order. We have started offering an option 
+> to inject our sidecar as the first container (previously, it was the last one), which has resulted in users running 
+> kubectl exec and getting the "wrong" container.
+
+
+### Goals
+
+
+- Provide a way to know which is the default container in pod for cli: kubectl
+
+
+### Non-Goals
+
+- If the cli is not kubectl, we don't determine which is the default container.
+
+
+## Proposal
+
+
+### Current Cli Behaviors
+
+Many Kubectl commands make requests to specific containers, `kubectl logs` and `kubectl exec` are most commonly
+used.
+
+There are currently 5 commands that allow specifying --container flag:
+- kubectl attach
+- kubectl cp
+- kubectl debug
+- kubectl exec
+- kubectl logs
+However, `kubectl debug` use this option to specify container name for debug container. 
+
+All commands that need to specify container name:
+
+- kubectl attach - Attach to a running container
+- kubectl cp - Copy files and directories to and from containers.
+- kubectl exec - Execute a command in a container
+
+These three commands above are similar. If omitted, the first container in the pod will be chosen.
+
+- kubectl logs - Print the logs for a container in a pod
+
+`kubectl logs` is a little different. Not default value for it. 
+However, it supports logs all containers with `--all-containers`.
+
+
+https://github.com/kubernetes/kubernetes/pull/87809 already add an annotation named
```

Comment:
I guess this part is pretty confuse, so let's rewrite this:

"
The following is the behavior of kubectl commands that can specify a container with --container:

* `kubectl attach`, `kubectl cp` and  `kubectl exec`

The three above have similar behavior: if --container flag is omitted, the first container of the Pod will be chosen

* `kubectl logs`
There's no default value. If a Pod have multiple containers, the operator needs to select which container to show the logs, or use the flag `--all-containers`

Also there's support for the annotation `kubectl.kubernetes.io/default-logs-container` in the Pod, that specified the default container to show the log

* `kubectl debug` 
Uses the argument XPTO (I don't know how it behaves here)

* 

"